### ISSUE TITLE
Move `self` out of `&mut self` in Iterator::next() implementation

### ISSUE THREAD

**Issue by mineichen** [rust-lang/rust#97396]
The compiler accepts calling other Iterator-Methods which require a owned self within Iterator methods which only have a &mut self.

```rust
struct MyIter;

impl Iterator for MyIter {
    type Item = usize;

    fn next(&mut self) -> Option<Self::Item> {
        self.count(); // Shouldn't be allowed!
        let _ = self.enumerate(); // Shouldn't be allowed!
        None
    }

    fn nth(&mut self, _n: usize) -> Option<Self::Item> {
        self.count(); // Shouldn't be allowed!
        let _ = self.enumerate(); // Shouldn't be allowed!
        None
    }

    fn size_hint(&self) -> (usize, Option<usize>) {
        //let x = self.count(); // Correctly  prohibits compilation
        //let x = self.enumerate(); // Correctly  prohibits compilation
        (0, Some(0))
    }
}

```

I expected to see a compiler error in this case.

Instead, it runs with undefined behavior...

This bug seems to affect just the Iterator-Trait. Custom traits are handled correctly.
```rust
trait NotAffected: Sized {
    fn test(&mut self) {
        //self.owned() // Correctly  prohibits compilation
    }
    fn owned(self) {}
}

impl NotAffected for MyIter {
    fn test(&mut self) {
        //self.owned() // Correctly  prohibits compilation
    }
}
```

Happens on nightly and Stable:
```
rustc 1.61.0 (fe5b13d68 2022-05-18)
binary: rustc
commit-hash: fe5b13d681f25ee6474be29d748c65adcd91f69e
commit-date: 2022-05-18
host: x86_64-pc-windows-msvc
release: 1.61.0
LLVM version: 14.0.0
```
```
rustc 1.63.0-nightly (76761db59 2022-05-24)
binary: rustc
commit-hash: 76761db5919b6d2e178b26947ad9b4fcb0ff0e7c
commit-date: 2022-05-24
host: x86_64-pc-windows-msvc
release: 1.63.0-nightly
LLVM version: 14.0.4
```

</details>

**Comment by SNCPlay42** [comment]
This is because an [`impl <I: Iterator> Iterator for &mut I`](https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html#impl-Iterator-14) exists. The `self.count()` and `self.enumerate()` calls are not consuming a `MyIter`, they are consuming a `&mut MyIter`.
